File: sp/go/pingpong/core/tdpnet.go
package tdpnet

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

type TDPnet struct {
	Conn *websocket.Conn
}

func NewTDPnet() *TDPnet {
	return &TDPnet{}
}

func (t *TDPnet) Connect(host string) error {
	dialer := websocket.DefaultDialer
	conn, _, err := dialer.Dial(host, nil)
	if err != nil {
		t.OnConnectionError(err.Error())
		return err
	}

	t.Conn = conn
	t.OnConnected(time.Now().UnixNano() / int64(time.Millisecond))

	// Start read loop in a background goroutine
	go func() {
		defer func() {
			t.Conn.Close()
			t.OnConnectionError("Connection closed")
		}()

		for {
			_, message, err := t.Conn.ReadMessage()
			if err != nil {
				t.OnError(err)
				break
			}

			// Try to parse as JSON first (matches handleEvent logic)
			var m struct {
				ID   interface{} `json:"id"`
				Data string      `json:"data"`
			}

			if err := json.Unmarshal(message, &m); err == nil && m.Data != "" {
				t.HandleEvent(m.ID, m.Data)
			} else {
				// If not valid JSON event, treat as raw packet data
				t.OnData(string(message))
			}
		}
	}()

	return nil
}

func (t *TDPnet) HandleEvent(id interface{}, data string) {
	packet, err := FromSP(data)
	if err != nil {
		return
	}

	subject := packet.Peering.Subject
	parts := strings.Split(subject, ":")
	category := parts[0]

	msgID := ""
	if len(parts) > 1 {
		msgID = parts[1]
	} else if id != nil {
		msgID = fmt.Sprintf("%v", id)
	}

	switch category {
	case "request":
		t.OnRequest(msgID, *packet)
	case "response":
		t.OnResponse(msgID, *packet)
	case "event":
		t.OnEvent(msgID, *packet)
	}
}

func (t *TDPnet) OnData(e string) {
	packet := FromString(e)
	if packet == nil {
		return
	}

	parts := strings.Split(packet.Peering.Subject, ":")
	category := parts[0]
	id := ""
	if len(parts) > 1 {
		id = parts[1]
	}

	switch category {
	case "request":
		t.OnRequest(id, *packet)
	case "response":
		t.OnResponse(id, *packet)
	case "event":
		t.OnEvent(id, *packet)
	}
}

// Lifecycle Hooks (Mocking JS override behavior)
func (t *TDPnet) OnConnected(timestamp int64)    {}
func (t *TDPnet) OnError(err error)              { log.Println("Error:", err) }
func (t *TDPnet) OnConnectionError(msg string)   { log.Println("Conn Error:", msg) }
func (t *TDPnet) OnRequest(id string, p Packet)  {}
func (t *TDPnet) OnResponse(id string, p Packet) {}
func (t *TDPnet) OnEvent(id string, p Packet)    {}

func (t *TDPnet) SendSignal(data string) error {
	if t.Conn != nil {
		return t.Conn.WriteMessage(websocket.TextMessage, []byte(data))
	}
	return fmt.Errorf("socket not connected")
}

func (t *TDPnet) SendData(data string) error {
	return t.SendSignal(data)
}

func (t *TDPnet) Request(from, to, id string, data interface{}) {
	jsonData, _ := json.Marshal(data)
	packetStr := CreateSignal(
		from,
		to,
		string(jsonData),
		"request:"+id,
		nil,
		0,
	)
	t.SendSignal(packetStr)
}

func (t *TDPnet) Response(packet Packet, id string, data interface{}) {
	jsonData, _ := json.Marshal(data)

	// Swap peering
	packet.Peering = Peering{
		Timestamp: uint64(time.Now().UnixNano() / 1000),
		From:      packet.Peering.To,
		To:        packet.Peering.From,
		Subject:   "response:" + id,
	}
	packet.Transformer.Data = string(jsonData)

	t.SendData(packet.ToString())
}

File: sp/python/pingpong/core/tdpnet.py
import asyncio
import websockets
import json
import time
from packet import Packet, Peering
from helper import Packeter

class TDPnet:
    def __init__(self):
        self._socket = None

    async def connect(self, host, ca_file=None):
        """
        Connects to the websocket host. 
        ca_file can be used to pass a path to a certificate authority file for SSL.
        """
        try:
            # If host starts with wss:// and ca_file is provided, you would configure ssl_context here
            async with websockets.connect(host) as websocket:
                self._socket = websocket
                self.on_connected(int(time.time() * 1000))
                
                # Main listen loop
                async for message in websocket:
                    try:
                        # Attempt to parse as JSON first (matches JS logic)
                        try:
                            m = json.loads(message)
                            self.handle_event(m.get('id'), m.get('data'))
                        except json.JSONDecodeError:
                            # If not JSON, treat as raw packet string
                            self.on_data(message)
                    except Exception as e:
                        self.on_error(e)
                        
                self.on_connection_error("Connection closed")
        except Exception as e:
            self.on_connection_error(e)
            raise e

    def on_connected(self, timestamp):
        """Override this method"""
        pass

    def on_error(self, err):
        """Override this method"""
        print(f"Error: {err}")

    def on_connection_error(self, err):
        """Override this method"""
        print(f"Connection Error: {err}")

    def handle_event(self, id, data):
        if not data:
            return
            
        packet = Packet.from_sp(data)
        subject = packet.peering.subject
        parts = subject.split(':')
        category = parts[0]
        msg_id = parts[1] if len(parts) > 1 else id

        if category == 'request':
            self.on_request(msg_id, packet)
        elif category == 'response':
            self.on_response(msg_id, packet)
        elif category == 'event':
            self.on_event(msg_id, packet)

    def on_data(self, e):
        packet = Packet.from_string(e)
        subject = packet.peering.subject
        parts = subject.split(':')
        category = parts[0]
        id = parts[1] if len(parts) > 1 else None

        if category == 'request':
            self.on_request(id, packet)
        elif category == 'response':
            self.on_response(id, packet)
        elif category == 'event':
            self.on_event(id, packet)

    def on_request(self, id, p):
        """Override this method"""
        pass

    def on_response(self, id, p):
        """Override this method"""
        pass

    def on_event(self, id, p):
        """Override this method"""
        pass

    async def send_signal(self, data):
        if self._socket:
            await self._socket.send(data)

    async def send_data(self, data):
        if self._socket:
            await self._socket.send(data)

    async def request(self, from_id, to_id, id, data):
        packet_str = Packeter.create_signal(
            from_id=from_id,
            to_id=to_id,
            subject=f"request:{id}",
            data=json.dumps(data)
        )
        await self.send_signal(packet_str)

    async def response(self, packet, id, data):
        # Swap peering for response: to -> from, from -> to
        packet.peering = Peering(
            from_id=packet.peering.to_id,
            to_id=packet.peering.from_id,
            subject=f"response:{id}"
        )
        packet.transformer.data = json.dumps(data)
        await self.send_data(str(packet))